// @elsassph/fast-source-map/lib/encoder.js
import { encodeVLQ } from './vlq';
var Encoder = /** @class */ (function () {
    function Encoder(writer) {
        this.writer = writer;
    }
    Encoder.prototype.separator = function () {
        this.writer.write(44); /* , */
    };
    Encoder.prototype.newline = function () {
        this.writer.write(59); /* ; */
    };
    Encoder.prototype.write5 = function (column, source, sourceLine, sourceColumn, name) {
        encodeVLQ(this.writer, column);
        encodeVLQ(this.writer, source);
        encodeVLQ(this.writer, sourceLine);
        encodeVLQ(this.writer, sourceColumn);
        encodeVLQ(this.writer, name);
    };
    Encoder.prototype.write4 = function (column, source, sourceLine, sourceColumn) {
        encodeVLQ(this.writer, column);
        encodeVLQ(this.writer, source);
        encodeVLQ(this.writer, sourceLine);
        encodeVLQ(this.writer, sourceColumn);
    };
    Encoder.prototype.write1 = function (column) {
        encodeVLQ(this.writer, column);
    };
    Object.defineProperty(Encoder.prototype, "length", {
        get: function () {
            return this.writer.length;
        },
        enumerable: true,
        configurable: true
    });
    return Encoder;
}());
export default Encoder;
